reasoner/planner/GreedyCostSearch.java

```diff
@@ -0,0 +1,102 @@
+import com.vaticle.typedb.core.traversal.TraversalEngine;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public abstract class GreedyCostSearch extends ReasonerPlanner {
+
+    public GreedyCostSearch(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
+        super(traversalEng, conceptMgr, logicMgr);
+    }
+
+    /* TODO:
+     * AnswerSize -> For each resolvable: Add the number of calls + the answers produced
+     * AnswersProduced? TraversalEngine tells us this based on constraints in the rule-body
+     * #calls? Product of the cardinality of each variable in the bounds (This number reduces as you evaluate a new resolvable which further constrains the variable - but min or less?)ssssssss
+     * Effect? Schedule stronger constraints first.
+     */
+    @Override
+    protected Plan<Resolvable<?>> planConjunction(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> inputBounds) {
+        Set<Identifier.Variable.Retrievable> bounds = new HashSet<>(inputBounds);
+        Set<Resolvable<?>> resolvables = compile(conjunction);
+
+        Set<Resolvable<?>> remaining = new HashSet<>(resolvables);
+        long cost = 0;
+        List<Resolvable<?>> orderedResolvables = new ArrayList<>();
+        Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies = dependencies(resolvables);
+
+        while (!remaining.isEmpty()) {
+            Optional<Resolvable<?>> nextResolvableOpt = remaining.stream()
+                    .filter(r -> dependenciesSatisfied(r, bounds, dependencies))
+                    .min(Comparator.comparing(r -> estimateCost(r, bounds)));
+
+            if (nextResolvableOpt.isEmpty()) {
+                nextResolvableOpt = remaining.stream()
+                        .filter(r -> !dependenciesSatisfied(r, bounds, dependencies) && !r.isNegated())
+                        .min(Comparator.comparing(r -> estimateCost(r, bounds)));
+            }
+            assert nextResolvableOpt.isPresent();
+            Resolvable<?> nextResolvable = nextResolvableOpt.get();
+            cost += estimateCost(nextResolvable, bounds); // TODO: Eliminate double work
+            orderedResolvables.add(nextResolvable);
+            remaining.remove(nextResolvable);
+            bounds.addAll(nextResolvable.retrieves());
+        }
+        assert resolvables.size() == orderedResolvables.size() && iterate(orderedResolvables).allMatch(r -> resolvables.contains(r));
+        return new Plan<>(orderedResolvables, cost);
+    }
+
+    abstract long estimateCost(Resolvable<?> resolvable, Set<Identifier.Variable.Retrievable> bounds);
+
+    public static class OldPlannerEmulator extends GreedyCostSearch {
```

Comment:
I assume that once we have the cost estimator implemented we'll get rid of this right?
I'd expect the `GreedyCostSearch` to merge into the `ReasonerPlanner` as well?

Reply:
Could. But for now I'd rather split the planner's interfacing & caching to be separate to the actual implementation of the search to be separate
I guess I could create a new class called PlanSearch and make it a member.

Reply:
it will be nested in the future